### PR TITLE
Python: Retain hostfn exception msg when thrown during ensemble.

### LINKED ISSUE #1067: "SWIG director method error" from HostFn exception inside Ensemble.
If an exception is thrown from a Python host function during a regular `CUDASimulation`'s execution, it is propagated back to python and a full error message and stack trace is received.

If the same occurs during a `CUDAEnsemble`, it is caught and handled by C/C++ so we only get this detail.

```
$ python boids_spatial3D.py
CUDAEnsemble completed 0 runs successfully!
There were a total of 4 errors.
Traceback (most recent call last):
  File "/home/rob/fgpu2/examples/python_rtc/boids_spatial3D_bounded/boids_spatial3D.py", line 426, in <module>
    cudaSimulation.simulate(rp);
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rob/fgpu2/build_py311/lib/Release/python/venv/lib/python3.11/site-packages/pyflamegpu/pyflamegpu.py", line 9147, in simulate
    return _pyflamegpu.CUDAEnsemble_simulate(self, plan)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pyflamegpu.pyflamegpu.FLAMEGPURuntimeException: (EnsembleError) /home/rob/fgpu2/src/flamegpu/simulation/CUDAEnsemble.cu(264): Run 2 failed on device 0, thread 2 with exception:
SWIG director method error.
```

According to the [SWIG documentation](https://www.swig.org/Doc4.0/SWIGDocumentation.html#Python_nn36), this is currently a limitation:

> This code will check the Python error state after each method call from a director into Python, and throw a C++ exception if an error occurred. This exception can be caught in C++ to implement an error handler. Currently no information about the Python error is stored in the Swig::DirectorMethodException object, but this will likely change in the future.

From digging around in the generated python, I found that the error handler which generates the C++ exception is passed a `PyObject*`, which presumably represents the exception.

```c++
void SwigDirector_HostFunction::run(flamegpu::HostAPI *arg0) {
  SWIG_PYTHON_THREAD_BEGIN_BLOCK;
  {
    swig::SwigVar_PyObject obj0;
    obj0 = SWIG_NewPointerObj(SWIG_as_voidptr(arg0), SWIGTYPE_p_flamegpu__HostAPI,  0 );
    if (!swig_get_self()) {
      Swig::DirectorException::raise("'self' uninitialized, maybe you forgot to call HostFunction.__init__.");
    }
#if defined(SWIG_PYTHON_DIRECTOR_VTABLE)
    const size_t swig_method_index = 0;
    const char *const swig_method_name = "run";
    PyObject *method = swig_get_method(swig_method_index, swig_method_name);
    swig::SwigVar_PyObject result = PyObject_CallFunctionObjArgs(method ,(PyObject *)obj0, NULL);
#else
    swig::SwigVar_PyObject swig_method_name = SWIG_Python_str_FromChar("run");
    swig::SwigVar_PyObject result = PyObject_CallMethodObjArgs(swig_get_self(), (PyObject *) swig_method_name ,(PyObject *)obj0, NULL);
#endif
    if (!result) {
      PyObject *error = PyErr_Occurred();
      {
        if (error != NULL) {
          throw Swig::DirectorMethodException();
        }
      }
    }
  }
  SWIG_PYTHON_THREAD_END_BLOCK;
}
```

I thought using `PyErr_Fetch()`, `PyErr_GetExcInfo()` or `PyException_GetTraceback()`/`PyException_GetContext()`/`PyException_GetCause()` I might be able to get the detail (https://docs.python.org/3/c-api/exceptions.html). But no luck. Fetch returns all `null` strings, and the rest cause the program to crash with no output (possible access violation or something if they're returning null ptrs).

Needs more investigation.

### PR DESCRIPTION
**Pull request by Robadob**
Tested:
* Windows/Debug/Python 3.9
* Linux/Release/Python 3.11

Closes #1067

### CODE REVIEW

**Comment by Robadob** [comment]
In both Windows and Linux CI failures, the error seems to be 

> pointer to incomplete class type "_frame" is not allowed

To resolve this locally I had to add the include Python's `frameobject.h`, so unclear why this is failing (the include isn't causing any error's itself). Possibly something that's changed between Python 3.9 and 3.11?

**Comment by Robadob** [comment]
Ensemble (fast failure) output with current version.

I think the exception message is printed twice as a consequence of `CUDAEnsemble` printing it, and Python then also printing it when it reaches Python.
```
(venv) C:\Users\Robadob\fgpu2\examples\python_rtc\boids_spatial3D_bounded>python boids_spatial3D.py
CUDAEnsemble completed 0 runs successfully!
There were a total of 4 errors.
C:\Users\Robadob\fgpu2\src\flamegpu\simulation\CUDAEnsemble.cu(264): Run 3 failed on device 0, thread 3 with exception:
SWIG director method error.
Exception was thrown during host function 'error_hostfn':
C:\Users\Robadob\fgpu2\examples\python_rtc\boids_spatial3D_bounded\boids_spatial3D.py(348):
Test generic python exception.


Traceback (most recent call last):
  File "C:\Users\Robadob\fgpu2\examples\python_rtc\boids_spatial3D_bounded\boids_spatial3D.py", line 503, in <module>
    cudaSimulation.simulate(rp);
  File "C:\Users\Robadob\fgpu2\build\lib\Debug\python\venv\lib\site-packages\pyflamegpu\pyflamegpu.py", line 9154, in simulate
    return _pyflamegpu.CUDAEnsemble_simulate(self, plan)
pyflamegpu.pyflamegpu.FLAMEGPURuntimeException: (EnsembleError) C:\Users\Robadob\fgpu2\src\flamegpu\simulation\CUDAEnsemble.cu(264): Run 3 failed on device 0, thread 3 with exception:
SWIG director method error.
Exception was thrown during host function 'error_hostfn':
C:\Users\Robadob\fgpu2\examples\python_rtc\boids_spatial3D_bounded\boids_spatial3D.py(348):
Test generic python exception.
```

**Comment by Robadob** [comment]
I may run this on Mav over the weekend with a py 3.11 env when I get time.

**Comment by Robadob** [comment]
Mav, Python 3.11, Release

`CUDASimulation`

```
(venv) (py311) rob@mavericks:~/fgpu2/examples/python_rtc/boids_spatial3D_bounded$ python boids_spatial3D.py
Traceback (most recent call last):
  File "/home/rob/fgpu2/examples/python_rtc/boids_spatial3D_bounded/boids_spatial3D.py", line 500, in <module>
    cudaSimulation.simulate();
    ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rob/fgpu2/build_py311/lib/Release/python/venv/lib/python3.11/site-packages/pyflamegpu/pyflamegpu.py", line 7566, in simulate
    return _pyflamegpu.CUDASimulation_simulate(self, *args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: SWIG director method error.
Exception was thrown during host function 'error_hostfn':
/home/rob/fgpu2/examples/python_rtc/boids_spatial3D_bounded/boids_spatial3D.py(350):
Test generic python exception.
```

`CUDAEnsemble`
```
(venv) (py311) rob@mavericks:~/fgpu2/examples/python_rtc/boids_spatial3D_bounded$ python boids_spatial3D.py
CUDAEnsemble completed 0 runs successfully!
There were a total of 4 errors.
Traceback (most recent call last):
  File "/home/rob/fgpu2/examples/python_rtc/boids_spatial3D_bounded/boids_spatial3D.py", line 426, in <module>
    cudaSimulation.simulate(rp);
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rob/fgpu2/build_py311/lib/Release/python/venv/lib/python3.11/site-packages/pyflamegpu/pyflamegpu.py", line 9150, in simulate
    return _pyflamegpu.CUDAEnsemble_simulate(self, plan)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pyflamegpu.pyflamegpu.FLAMEGPURuntimeException: (EnsembleError) /home/rob/fgpu2/src/flamegpu/simulation/CUDAEnsemble.cu(264): Run 2 failed on device 0, thread 2 with exception:
SWIG director method error.
Exception was thrown during host function 'error_hostfn':
/home/rob/fgpu2/examples/python_rtc/boids_spatial3D_bounded/boids_spatial3D.py(350):
Test generic python exception.
```

Interestingly, the `CUDAEnsemble` doesn't double print the message on Mav.

-------------

I think this is now ready fully functional, with the only remaining change/s being if we want to refine the error message or switch out the exception type (e.g. from the stock `Swig::DirectorMethodException`).

One obvious improvement would be to remove the trailing `\n` on the error message, as it currently leaves 2 blank lines when the message is printed. Looks a bit dumb.

File: swig/python/flamegpu.i [hunk 1]
```diff
@@ -313,12 +313,61 @@ class FLAMEGPURuntimeException : public std::exception {
     std::string str_str;
 };
 %}
-
+%inline %{
+    #include <sstream>
+    // Python include for stacktrace frames pre Python 3.11
+    #if PY_VERSION_HEX < 0x030900B1
```
Comment (ptheywood) [inline swig/python/flamegpu.i:319]: The macro is actually checking LT 3.9.0b1, not 3.11 (though this is the version that is 3.11 friendly) so the comment is maybe not the most clear but not worth the change.

**Review by ptheywood — APPROVED** [review]
Current PR CI is only checking python 3.11, so 3.7, 3.8 and 3.10 are untested. Should be fine and will be caught during the next release if it is an error.

It builds with 3.10 for me under linux, but as this doesnt' add a test I've got no way to check the exception handling from within an ensemble without writing an example. 

Extra newlines don't matter to be honest, it'll only occur when an exception is triggered in the ensemble, in whcih case it needs to grab attention anyway.

Looks fine, but without checking it works on all python versions we release we should do another Rc next.

**Review by ptheywood — CHANGES_REQUESTED** [review]
Scratch the last approval / in ability to test this. 

`test_host_agent_creation.py` includes a few tests which intentionally trigger exceptions from python host functions. 

These tests now fail. 

Have you ran pytest? 

```
pytest ../tests/python/runtime/agent/test_host_agent_creation.py 
```

Ouput for th final of the 16 failures:

```
________________________________________________________________ HostAgentCreationTest.test_reserved_name_array ________________________________________________________________

self = <test_host_agent_creation.HostAgentCreationTest testMethod=test_reserved_name_array>

    def test_reserved_name_array(self):
        model = pyflamegpu.ModelDescription("model")
        model.newAgent("agent_name")
        func = reserved_name_step_array()
        model.addStepFunction(func)
        sim = pyflamegpu.CUDASimulation(model)
        with pytest.raises (pyflamegpu.FLAMEGPURuntimeException) as e:
>           sim.step()

../tests/python/runtime/agent/test_host_agent_creation.py:794: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pyflamegpu.pyflamegpu.CUDASimulation; proxy of <Swig Object of type 'flamegpu::CUDASimulation *' at 0x7ff3908f4990> >

    def step(self):
        r"""
        Steps the simulation once
        :rtype: boolean
        :return: False if an exit condition was triggered
        """
>       return _pyflamegpu.CUDASimulation_step(self)
E       RuntimeError: SWIG director method error. 
E       FLAMEGPURuntimeException was thrown during host function 'reserved_name_step_array':
E       /home/ptheywood/code/flamegpu/FLAMEGPU2/tests/python/runtime/agent/test_host_agent_creation.py(198): 
E       (ReservedName) /home/ptheywood/code/flamegpu/FLAMEGPU2/include/flamegpu/runtime/agent/HostNewAgentAPI.h(411): Agent variable names cannot begin with '_', this is reserved for internal usage, in HostNewAgentAPI::setVariable().

lib/Release/python/venv/lib/python3.10/site-packages/pyflamegpu/pyflamegpu.py:7542: RuntimeError
```
```

File: swig/python/flamegpu.i (hunk 1 above)
Comment (Robadob) [reply swig/python/flamegpu.i:319]: Got that hex thing from the py 3.11 changelog.

**Comment by Robadob** [comment]
> Have you ran pytest?

No, I'll get to that this evening. I presume it's checking the exception type and/or message.

**Comment by Robadob** [comment]
Have resolved the python test failures, I was not restoring the Python exception state.

Likewise, the previously tested behaviour added with this PR within `CUDAEnsemble` is still working.

**Review by mondus — APPROVED** [review]

**Review by mondus — APPROVED** [review]

**Pull request merged**